doc/cookbook/using-free-client/UsingFreeClient.lhs

```diff
@@ -0,0 +1,173 @@
+# Using Free Client (for tests)
+
+Someone asked on IRC about getting `Request` & `Response` of what
+`servant-client` uses, for testing purposes.  My response: to ad-hoc extend
+`servant-client` (as for tests), use `Servant.Client.Free`. This recipe is an
+evidence.
+
+First the module header, but this time We'll comment the imports.
+
+```haskell
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+module Main (main) where
+```
+
+We will primarily use `Servant.Client.Free`, it doesn't re-export anything
+from `free` package, so we need to import it as well.
+
+```haskell
+import Control.Monad.Free
+import Servant.Client.Free
+```
+
+Also we'll use `servant-client` internals, which uses `http-client`,
+so let's import them *qualified*
+
+```haskell
+import qualified Servant.Client.Internal.HttpClient as I
+import qualified Network.HTTP.Client                as HTTP
+```
+
+The rest of the imports are for a server we implement here for completeness.
+
+```haskell
+import Servant
+import Network.Wai.Handler.Warp (run)
+import System.Environment       (getArgs)
+```
+
+## API & Main
+
+We'll work with a very simple API:
+
+```haskell
+type API = "square" :> Capture "n" Int :> Get '[JSON] Int
+
+api :: Proxy API
+api = Proxy
+```
+
+Next we implement a `main`. If passed `server` it will run `server`, if
+`client` a small `test` (to be defined next) will be run. This should be pretty
+straigh-forward:
+
+```haskell
+main :: IO ()
+main = do
+    args <- getArgs
+    case args of
+        ("server":_) -> do
+            putStrLn "Starting cookbook-using-free-client at http://localhost:8000"
+            run 8000 $ serve api $ \n -> return (n * n)
+        ("client":_) ->
+            test
+        _ -> do
+            putStrLn "Try:"
+            putStrLn "cabal new-run cookbook-using-free-client server"
+            putStrLn "cabal new-run cookbook-using-free-client client"
+```
+
+## Test
+
+In the actual test, we'll use a `Servant.Client.Free` client.
+Cecause we have a single endpoint API, we'll get a single client function:
+
+```haskell
+fcli :: Int -> Free ClientF Int
+fcli = client api
+```
+
+Next, we can write our small test. We'll pass a value to `fcli` and inspect
```

Comment:
I'd probably try to add a small paragraph here saying that such clients are basically "client functions without a backend", or where the backend has been abstracted out. And that `ClientF` is precisely a functor that represents the operations we need from an http backend. So if we are to emulate one or augment what it does, it will be by interpreting all those operations the way we want to. This also means we get access to the requests and responses and can do anything we want we them as they are produced or consumed, respectively. 

Just as a little intro to the `case ... of` block below. To give it a bit of context.

Reply:
If you add *client functions without a backend", than you could mention that using `Free` is (probably?) a little easier than implementing own `RunClient m` instance, though it requires to implement only 3 functions too (atm, maybe we'll go do two, let's see). `Free` however is nice especially for single-stepping will do later.

**But** if you only, say need to remember `Request`s made by client, you can write istance for `WriterT [Request] ClientM`. and inspect them afterwards.

Reply:
them = Requests